README.md
# YII2 Việt Nam Validator
**Yii2 Extension hổ trợ bạn kiểm tra dữ liệu đặc thù trong nước ta.**

[![Latest Stable Version](https://poser.pugx.org/yiiviet/yii2-validator/v/stable)](https://packagist.org/packages/yiiviet/yii2-validator)
[![Total Downloads](https://poser.pugx.org/yiiviet/yii2-validator/downloads)](https://packagist.org/packages/yiiviet/yii2-validator)
[![Build Status](https://travis-ci.org/yiiviet/yii2-validator.svg?branch=master)](https://travis-ci.org/yiiviet/yii2-validator)
[![Code Coverage](https://scrutinizer-ci.com/g/yiiviet/yii2-validator/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/yiiviet/yii2-validator/?branch=master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/yiiviet/yii2-validator/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/yiiviet/yii2-validator/?branch=master)
[![Yii2](https://img.shields.io/badge/Powered_by-Yii_Framework-green.svg?style=flat)](http://www.yiiframework.com/)

Khi bạn xây dựng hệ thống trong nước thì chắc chắn rằng việc xây dựng các `validator` để kiểm tra dữ liệu
đặc thù như: số điện thoại, chứng minh thư, ip Việt Nam, domain Việt Nam... sẽ thường xuyên xảy ra. 
Chính vì vậy extension này được xây dựng nên để cung cấp các `validator` phổ biến nhất về các dữ liệu
cơ bản tại nước ta.


Các kiểu dữ liệu hiện được hổ trợ:

* Số điện thoại bao gồm số di động và số bàn.
* Chứng minh thư và thẻ căn cước.
* IP Việt Nam
* Tên miền Việt Nam


## Cài đặt

Cài đặt thông qua `composer` nếu như đó là một khái niệm mới với bạn xin click vào 
[đây](http://getcomposer.org/download/) để tìm hiểu nó.

```sh
composer require "yiiviet/yii2-validator"
```

hoặc thêm

```json
"yiiviet/yii2-validator": "*"
```

vào phần `require` trong file composer.json.

## Cách sử dụng cơ bản

Về cách sử dụng thì bạn khai báo nó vào `rules` của `model`, `AR` như các `validator`
thuần của `yii2`


Ví dụ:

```php

class Info extends Model {

    public function rules() {
    
        return [
            ['IP', 'ipvn'],
            ['SDT', 'telnumvn'],
            ['CMND', 'idnumvn'],
            ['domain', 'domainvn']
        ];
    
    }
}
```

Bảng chú thích tên các `validator`:

| Validator | Chú thích | Cách dùng |
| :--------: | ---------- | -------- |
| telnumvn | Dùng để kiểm tra số điện thoại Việt Nam (viettel, vina, vsat, beeline, mobi, vietnammobi, số bàn...) | [Nhấn vào đây](#telnumvn)
| idnumvn | Dùng để kiểm tra chứng minh thư, thẻ căn cước (chứng minh 9 số, 12 số, thẻ căn cước) | [Nhấn vào đây](#idnumvn)
| ipvn | Dùng để kiểm tra IP Việt Nam | [Nhấn vào đây](#ipvn)
| domainvn | Dùng để kiểm tra tên miền Việt Nam (là tên miền hợp lệ và kết thúc bằng `.vn`) | [Nhấn vào đây](#domainvn)

### `telnumvn`

Khai báo rules:

```php

    return [
        ['SDT', 'telnumvn', 'message' => '{attribute} không phải là số điện thoại Việt Nam'],
    ];

```

Chỉ muốn kiểm tra số di động loại bỏ số bàn

```php

    return [
        ['SDT', 'telnumvn', 'exceptTelco' => ['landLine']],
    ];

```

Chỉ muốn kiểm tra số `viettel`

```php

    return [
        ['SDT', 'telnumvn', 'onlyTelco' => ['viettel']],
    ];

```

Chỉ muốn kiểm tra số `viettel`, `mobi`, `vina`

```php

    return [
        ['SDT', 'telnumvn', 'onlyTelco' => ['viettel', 'mobiFone', 'vinaPhone']],
    ];

```

Danh sách tên các telco (nhà mạng)

| Telco | Đại diện | 
| :--------: | ---------- |
| landLine | Số điện thoại bàn | 
| viettel | Viettel |
| vinaPhone | Vinaphone | 
| mobiFone | Mobifone |
| vietNamMobile | Viet Nam Mobile |
| gMobile | G Mobile |
| indoChina | IndoChina |

### `idnumvn`

Khai báo rules:

```php

    return [
        ['CMND', 'idnumvn', 'message' => '{attribute} không phải là chứng minh thư Việt Nam'],
    ];

```

Chỉ muốn kiểm tra số chứng minh (9 và 12 số) thư loại bỏ thẻ căn cước

```php

    return [
        ['SDT', 'telnumvn', 'onlyId' => true],
    ];

```

### `ipvn`

Khai báo rules:

```php

    return [
        ['IP', 'ipvn', 'message' => '{attribute} phải là IP Việt Nam!'],
    ];

```

Do `validator` này kế thừa `IpValidator` của `Yii2` nên tất cả các tham số thiết lập nâng cao
bạn có thể kham khảo thêm tại [đây](https://www.yiiframework.com/doc/guide/2.0/en/tutorial-core-validators#ip), còn không bạn chỉ cần khai báo như trên là đủ.

### `domainvn`


Khai báo rules:

```php

    return [
        ['domain', 'domainvn', 'message' => '{attribute} phải là tên miền Việt Nam!'],
    ];

```

Do `validator` này kế thừa `DomainValidator` của `Dmitry Kulikov` nên tất cả các tham số thiết lập nâng cao
bạn có thể kham khảo thêm tại [đây](https://github.com/dmitry-kulikov/yii2-domain-validator), còn không bạn chỉ cần khai báo như trên là đủ.

## Dành cho nhà phát triển

Nếu như bạn cảm thấy các `validators` bên trên vẫn chưa đủ đối với thị trường trong nước và bạn muốn
đóng góp để phát triển chung, chúng tôi rất hoan nghênh! Hãy tạo các `issue` để đóng góp ý tưởng cho
phiên bản kế tiếp hoặc tạo `PR` để đóng góp thêm các `validator` còn thiếu sót. Cảm ơn!


CHANGELOG.md
# 1.0.3 

Lỗi #1: không kiểm tra được đầu số `099` của `gMobile` (vuongxuongminh)

# 1.0.2

Tính năng: cập nhật `pattern` của `yiiviet\validator\TelNumValidator` (vuongxuongminh)


tests/TestCase.php
<?php
/**
 * @link https://github.com/yiiviet/yii2-payment
 * @copyright Copyright (c) 2017 Yii2VN
 * @license [New BSD License](http://www.opensource.org/licenses/bsd-license.php)
 */

namespace yiiviet\tests\unit\validator;

use Yii;

use yii\helpers\ArrayHelper;
use yiiviet\validator\Bootstrap;

use PHPUnit\Framework\TestCase as BaseTestCase;

/**
 * Class TestCase
 *
 * @author Vuong Minh <vuongxuongminh@gmail.com>
 * @since 1.0
 */
class TestCase extends BaseTestCase
{


    public function setUp()
    {
        $this->mockApplication();

        parent::setUp();
    }

    public function tearDown()
    {
        $this->destroyApplication();

        parent::tearDown();
    }

    protected function mockApplication($config = [], $appClass = '\yii\console\Application')
    {
        new $appClass(ArrayHelper::merge([
            'id' => 'testapp',
            'basePath' => __DIR__,
            'bootstrap' => [
                Bootstrap::class
            ],
            'vendorPath' => dirname(dirname(__DIR__)) . '/vendor',
        ], $config));
    }

    /**
     * Destroys application in Yii::$app by setting it to null.
     */
    protected function destroyApplication()
    {
        Yii::$app = null;
    }


}


tests/bootstrap.php
<?php
// ensure we get report on all possible php errors
error_reporting(-1);

define('YII_ENABLE_ERROR_HANDLER', false);
define('YII_DEBUG', true);

require_once(__DIR__ . '/../vendor/autoload.php');
require_once(__DIR__ . '/../vendor/yiisoft/yii2/Yii.php');
require_once(__DIR__ . '/compatibility.php');

Yii::setAlias('@yiiviet/tests/unit/validator', __DIR__);
Yii::setAlias('@yiiviet/validator', dirname(__DIR__) . '/src');


tests/IpValidatorTest.php
<?php
/**
 * @link https://github.com/yiiviet/yii2-validator
 * @copyright Copyright (c) 2018 Yii Viet
 * @license [New BSD License](http://www.opensource.org/licenses/bsd-license.php)
 */

namespace yiiviet\tests\unit\validator;

use yii\base\DynamicModel;

/**
 * Lớp IpValidatorTest
 *
 * @author Vuong Minh <vuongxuongminh@gmail.com>
 * @since 1.0
 */
class IpValidatorTest extends TestCase
{

    /**
     * @dataProvider validIpProvider
     */
    public function testValidIp($ip)
    {
        $model = DynamicModel::validateData(['ip' => $ip], [
            ['ip', 'ipvn']
        ]);

        $this->assertFalse($model->hasErrors());
    }

    /**
     * @dataProvider invalidIpProvider
     */
    public function testInvalidIp($ip)
    {
        $model = DynamicModel::validateData(['ip' => $ip], [
            ['ip', 'ipvn']
        ]);

        $this->assertTrue($model->hasErrors());
    }


    public function validIpProvider()
    {
        return [
            ['113.161.173.10'],
            ['171.255.199.129'],
            ['118.70.187.126'],
            ['115.78.225.211'],
            ['2405:4800:102:1::3'],
            ['2001:df0:66:40::16']
        ];
    }

    public function invalidIpProvider()
    {
        return [
            ['217.24.254.30'],
            ['82.114.86.1'],
            ['200.114.80.9'],
            ['190.123.83.251'],
            ['190.123.83.251'],
            ['190.123.83.251'],
            ['139.199.201.249'],
            ['52.80.73.123'],
            ['2a03:2880:f11f:83:face:b00c::25de'],
            ['2a00:1450:4007:809::200e']
        ];
    }
}


src/Bootstrap.php
<?php
/**
 * @link https://github.com/yiiviet/yii2-validator
 * @copyright Copyright (c) 2018 Yii Viet
 * @license [New BSD License](http://www.opensource.org/licenses/bsd-license.php)
 */

namespace yiiviet\validator;

use yii\base\BootstrapInterface;
use yii\validators\Validator;

/**
 * Lớp Bootstrap dùng để khai báo các validator vào built-in của YII2, từ đó tạo nên sự đơn giản hóa khi sử dụng.
 *
 * @author Vuong Minh <vuongxuongminh@gmail.com>
 * @since 1.0
 */
class Bootstrap implements BootstrapInterface
{

    /**
     * @inheritdoc
     */
    public function bootstrap($app)
    {
        Validator::$builtInValidators = array_merge([
            'ipvn' => IpValidator::class,
            'idnumvn' => IdNumValidator::class,
            'telnumvn' => TelNumValidator::class,
            'domainvn' => DomainValidator::class
        ], Validator::$builtInValidators);
    }

}


src/IpValidator.php
<?php
/**
 * @link https://github.com/yiiviet/yii2-validator
 * @copyright Copyright (c) 2018 Yii Viet
 * @license [New BSD License](http://www.opensource.org/licenses/bsd-license.php)
 */

namespace yiiviet\validator;

use yii\validators\IpValidator as BaseIpValidator;

/**
 * Lớp IpValidator dùng để kiểm tra `ip` Việt Nam.
 * Ví dụ khai báo kiểm tra ip trong model
 *
 * ```php
 *      public function rules() {
 *          return [
 *              ['ipAttr', 'ipvn', 'message' => 'Ip phải là ip Việt Nam!']
 *          ];
 *      }
 * ```
 *
 * @author Vuong Minh <vuongxuongminh@gmail.com>
 * @since 1.0
 */
class IpValidator extends BaseIpValidator
{

    /**
     * @var array Thuộc tính hổ trợ cache range ip Việt Nam.
     */
    private static $_ipRanges = [];

    /**
     * @throws \yii\base\InvalidConfigException
     */
    public function init()
    {
        parent::init();

        if (!empty(self::$_ipRanges)) {
            $this->setRanges(self::$_ipRanges);
        } else {
            $this->setRanges(self::$_ipRanges = require(__DIR__ . '/resource/ip-ranges.php'));
        }
    }


}


src/IdNumValidator.php
<?php
/**
 * @link https://github.com/yiiviet/yii2-validator
 * @copyright Copyright (c) 2018 Yii Viet
 * @license [New BSD License](http://www.opensource.org/licenses/bsd-license.php)
 */

namespace yiiviet\validator;

use yii\validators\RegularExpressionValidator;

/**
 * Lớp IdNumValidator dùng để kiểm tra số chứng minh thư có hợp lệ hay không.
 * Ví dụ khai báo kiểm tra số chứng minh thư trong model
 *
 * ```php
 *      public function rules() {
 *          return [
 *              ['idAttr', 'idnumvn', 'message' => 'Chứng minh nhân dân không hợp lệ.']
 *          ];
 *      }
 * ```
 *
 * Nếu bạn chấp nhận luôn thẻ căn cước vừa mới ban hành thì:
 *
 * ```php
 *      public function rules() {
 *          return [
 *              ['idAttr', 'idnumvn', 'message' => 'Chứng minh nhân dân hoặc thẻ căn cước không hợp lệ.', 'onlyId' => false]
 *          ];
 *      }
 * ```
 *
 * @author Vuong Minh <vuongxuongminh@gmail.com>
 * @since 1.0
 */
class IdNumValidator extends RegularExpressionValidator
{

    /**
     * @var bool Thiết lập nếu như bạn chỉ muốn kiểm tra chứng minh thư (không bao gồm thẻ căn cước).
     */
    public $onlyId = false;

    /**
     * @var string Pattern của chứng minh thư cũ.
     */
    public $id = '^((0[0-8]|1[0-9]|2[0-9]|3[0-8])\d{7})|((09[015]|23[01]|245|28[015])\d{6})$';

    /**
     * @var string Pattern của chứng minh thư mới.
     */
    public $idNew = '^((0[0-8]|1[0-9]|2[0-9]|3[0-8])\d{10})|((09[015]|23[01]|245|28[015])\d{9})$';

    /**
     * @var string Pattern của căn cước công dân.
     */
    public $cId = '^(0[012468]|1[0124579]|2[02]|2[4-7]|3[01]|3[3-8]|4[0245689]|5[12468]|6[024678]|7[024579]|8[0234679]|9[1-6])\d{10}$';

    /**
     * {@inheritdoc}
     * @throws \yii\base\InvalidConfigException
     */
    public function init()
    {
        $pattern = [$this->id, $this->idNew];

        if (!$this->onlyId) {
            $pattern[] = $this->cId;
        }

        $this->pattern = "/(" . implode(")|(", $pattern) . ")/";

        parent::init();
    }

}


src/DomainValidator.php
<?php
/**
 * @link https://github.com/yiiviet/yii2-validator
 * @copyright Copyright (c) 2018 Yii Viet
 * @license [New BSD License](http://www.opensource.org/licenses/bsd-license.php)
 */

namespace yiiviet\validator;

use kdn\yii2\validators\DomainValidator as BaseDomainValidator;

/**
 * Lớp DomainValidator dùng để kiểm tra tên miền có phải là của Việt Nam hay không.
 * Ví dụ khai báo kiểm tra domain Việt Nam trong `model`
 *
 * ```php
 *      public function rules() {
 *          return [
 *              ['domainAttr', 'domainvn', 'message' => 'Tên miền không phải miền VN!']
 *          ];
 *      }
 * ```
 *
 * @author Vuong Minh <vuongxuongminh@gmail.com>
 * @since 1.0
 */
class DomainValidator extends BaseDomainValidator
{

    /**
     * @inheritdoc
     * @throws \yii\base\InvalidConfigException
     */
    public function init()
    {
        parent::init();

        if (empty($this->message)) {
            $this->message = '{attribute} is invalid `.vn` domain!';
        }
    }

    /**
     * @inheritdoc
     * @throws \yii\base\NotSupportedException
     */
    public function validateValue($value)
    {
        if ($result = parent::validateValue($value)) {
            return $result;
        } elseif (!preg_match('/\.vn$/', $value)) {
            return [$this->message, []];
        } else {
            return null;
        }
    }

}


src/TelNumValidator.php
<?php
/**
 * @link https://github.com/yiiviet/yii2-validator
 * @copyright Copyright (c) 2018 Yii Viet
 * @license [New BSD License](http://www.opensource.org/licenses/bsd-license.php)
 */

namespace yiiviet\validator;

use yii\base\InvalidConfigException;
use yii\validators\RegularExpressionValidator;

/**
 * Lớp TelNumValidator dùng để kiểm tra số điện thoại trong nước có hợp lệ hay không. Nó được phát triển theo tài liệu wiki
 * link: https://vi.wikipedia.org/wiki/M%C3%A3_%C4%91i%E1%BB%87n_tho%E1%BA%A1i_Vi%E1%BB%87t_Nam
 *
 * Ví dụ khai báo kiểm tra số điện thoại hợp lệ trong `model`
 *
 * ```php
 *      public function rules() {
 *          return [
 *              ['telAttr', 'telnumvn', 'message' => 'Số điện thoại phải là số trong nước']
 *          ];
 *      }
 * ```
 *
 * Chỉ kiểm tra số di động (bỏ số bàn)
 *
 * ```php
 *      public function rules() {
 *          return [
 *              ['telAttr', 'telnumvn', 'message' => 'Số điện thoại phải là số trong nước', 'exceptTelco' => 'landLine']
 *          ];
 *      }
 * ```
 *
 * Chỉ kiểm tra số bàn (bỏ số di động)
 *
 * ```php
 *      public function rules() {
 *          return [
 *              ['telAttr', 'telnumvn', 'message' => 'Số điện thoại phải là số trong nước', 'onlyTelco' => 'landLine']
 *          ];
 *      }
 * ```
 *
 * danh sách các telco (nhà mạng) hổ trợ: mobiFone, vinaPhone, viettel, vietNamMobile, gMobile, beeline, vsat, indoChina, landLine.
 *
 * Sử dụng trong DynamicModel tương tự.
 *
 * @author Vuong Minh <vuongxuongminh@gmail.com>
 * @since 1.0
 */
class TelNumValidator extends RegularExpressionValidator
{

    /**
     * @var array Mảng chứa tên các telco (nhà mạng). Các nhà mạng nằm trong mảng sẽ là kiểu ràng buộc của dữ liệu.
     * Lưu ý nếu như tên telco cũng nằm ở mảng `exceptTelco` thì mặc định telco đó sẽ bị loại bỏ.
     *
     * Nếu mảng là rổng động nghĩa với dữ liệu kiểm tra chỉ cần là số điện thoại Việt Nam.
     *
     * @see exceptTelco
     */
    public $onlyTelco = [];

    /**
     * @var array Mảng chứa tên các telco (nhà mạng). Các nhà mạng nằm trong mảng sẽ được bỏ qua (không kiểm tra).
     * @see onlyTelco
     */
    public $exceptTelco = [];

    /**
     * @var string Pattern kiểm tra số viettel.
     */
    public $viettel = '^(\+?84|0)?(3[2-9]|86|9[6-8])\d{7}$';

    /**
     * @var string Pattern kiểm tra số vinaphone.
     */
    public $vinaPhone = '^(\+?84|0)?(8[1-5]|88|9[14])\d{7}$';

    /**
     * @var string Pattern kiểm tra số mobiphone.
     */
    public $mobiFone = '^(\+?84|0)?(70|7[6-9]|89|9[03])\d{7}$';

    /**
     * @var string Pattern kiểm tra số vnmobi.
     */
    public $vietNamMobile = '^(\+?84|0)?(5[68]|92)[\d]{7}$';

    /**
     * @var string Pattern kiểm tra số gmobile.
     */
    public $gMobile = '^(\+?84|0)?([59]9|95)[\d]{7}$';

    /**
     * @var string Pattern kiểm tra số indochina.
     */
    public $indoChina = '^(\+?84|0)?87[\d]{7}$';

    /**
     * @var string Pattern kiểm tra số điện thoại bàn.
     */
    public $landLine = '^(\+?84|0)?(((20[3-9]|21[0-6]|21[89]|22[0-2]|22[5-9]|23[2-9]|24[2-5]|248|25[12]|25[4-9]|26[0-3]|27[0-7]|28[2-5]|29([0-4]|[67])|299)\d{7})|((246[236]|247[13]|286[23]|287[13])\d{6}))$';

    /**
     * @var bool|string Thiết lập kiểu `format` di động thêm '0' sau khi thực thi kiểm tra hoàn tất (dữ liệu attr hợp lệ).
     * Nó dùng để chuẩn hóa dữ liệu cho bạn. Mặc định `FALSE` sẽ không can thiệp đến dữ liệu của bạn.
     */
    public $mobileFormat = false;

    /**
     * @inheritdoc
     * @throws InvalidConfigException
     */
    public function init()
    {
        $this->onlyTelco = (array)$this->onlyTelco;
        $this->exceptTelco = (array)$this->exceptTelco;

        $pattern = [];
        foreach (['viettel', 'vinaPhone', 'mobiFone', 'vietNamMobile', 'indoChina', 'gMobile', 'landLine'] as $telco) {
            if ($this->isUse($telco)) {
                $pattern[] = $this->{$telco};
            }
        }

        if (!empty($pattern)) {
            $this->pattern = "~(" . implode(")|(", $pattern) . ")~";
        } else {
            throw new InvalidConfigException('Your telco setup is not valid!');
        }

        parent::init();
    }


    /**
     * Phương thức kiểm tra telco (nhà mạng) có phải là thành phần cần kiểm tra hay không.
     * Nó được trích từ [yii\base\ActionFilter::isActive()].
     *
     * @param string $telco Tên nhà mạng cần kiểm tra
     * @return bool Trả về `TRUE` nếu như telco la thành phần kiểm tra.
     */
    protected function isUse($telco)
    {
        if (empty($this->onlyTelco)) {
            $only = true;
        } else {
            $only = false;
            foreach ($this->onlyTelco as $expect) {
                if ($telco === $expect) {
                    $only = true;
                    break;
                }
            }
        }

        $except = false;
        foreach ($this->exceptTelco as $expect) {
            if ($telco === $expect) {
                $except = true;
                break;
            }
        }

        return $only && !$except;
    }

    /**
     * @inheritdoc
     */
    public function validateAttribute($model, $attribute)
    {
        if (($result = parent::validateAttribute($model, $attribute)) === null) {
            if ($this->mobileFormat) {
                $model->{$attribute} = preg_replace('/^(\+?84|0)?(\d+)$/', '0$2', $model->{$attribute});
            }
        }

        return $result;
    }

}
